### PR TITLE
ADS-6054: Adding more prebid events

### PR DESCRIPTION
**Pull request by AritraSengupta**
Ticket: https://mavencorp.atlassian.net/browse/ADS-6054

### CODE REVIEW

**Comment by AritraSengupta** [comment]
Tested this with @partha0solapurkar, we are able to get most of the stuff requested in the PR. Few things to note:
1. Floor price will be 0 if the it is not set on the bid level or the creative is not banner or video. Native ads will reflect a floor price 0.
2. I have added an `error` bid_status, to capture rejected bids by prebid.

**Comment by partha0solapurkar** [comment]
> I have added an `error` bid_status, to capture rejected bids by prebid.

This is a great idea! @martiantim in my local testing for rejected bids, the only reason for rejection I saw was `bid does not meet price floor` and a bid amount and a bid response time were present. What do you think about collecting that data for the rejected bids along with the reason for rejection? 

Here are a few pros:

1. It would be useful if we suspect the floors are too high and want to analyze whether we have something to gain by lowering the floor. 

2. We can just reuse `bid_amountss` and `bid_response_timess` to record the numbers for the rejected bids in addition to the non-rejected bids. Currently they are 0 and null respectively for the rejected bids, so the only thing to keep in mind would be to filter on bid_status for some purposes. 

3. We don't need to add the reason for rejection to BQ as it is, but if we start collecting it, we can dig into the other reasons if there are any and later decide whether we want to add a summarized reason to BQ. 

I couldn't think of any cons. Can you think of any?

**Comment by AritraSengupta** [comment]
@partha0solapurkar specifically for the bidRejected ones, do you think it will be helpful if we did the following:
1. send `error: {reason}` instead of just `error`
2. in the `bid_amounts`, I can send the cpm value from the bid rejected if it exists, so we will know the amount they bid at
@martiantim  thoughts?

**Comment by martiantim** [comment]
Sounds great

Tim Martin

Vice President, Head of Engineering


Phone: +1 (510) 260-4175

Email: ***@***.*** ***@***.***>

Web: thearenagroup.net

**Comment by partha0solapurkar** [comment]
@AritraSengupta Yes, that sounds good to me. I'll dig into the error reasons first and then decide whether we need to send (a condensed version of) them to BigQuery, in case some of the reviewers are concerned about the size of the reason and the implications for storage costs.

File: modules/mavenDistributionAnalyticsAdapter.js [hunk 1]
```diff
@@ -168,9 +168,9 @@ const getBidStatusAmtsAndResponseTime = (key, bidRequest, args) => {
   // Config decribing the various bidStatus, bidAmount and bidResponseTime
   // values for each bid condition
   const BID_STATUS_MAP = {
-    bidsRejected: () => ({
-      bidStatus: 'error',
-      bidAmount: 0,
+    bidsRejected: (bid) => ({
+      bidStatus: `error: ${bid.rejectionReason || 'generic'}`,
+      bidAmount: bid.cpm || 0,
```
Comment (partha0solapurkar) [inline modules/mavenDistributionAnalyticsAdapter.js]: `Math.round((bid.cpm || 0) * 1000)`

**Review by partha0solapurkar — CHANGES_REQUESTED** [review]

File: modules/mavenDistributionAnalyticsAdapter.js [hunk 2]
```diff
@@ -122,13 +109,121 @@ export function summarizeAuctionInit(args, adapterConfig) {
...
+}
+
+const getBidStatusAmtsAndResponseTime = (key, bidRequest, args) => {
+  // Config decribing the various bidStatus, bidAmount and bidResponseTime
+  // values for each bid condition
+  const BID_STATUS_MAP = {
+    bidsRejected: (bid) => ({
+      bidStatus: `error: ${bid.rejectionReason || 'generic'}`,
+      bidAmount: bid.cpm || 0,
+      bidResponseTime: null,
```
Comment (partha0solapurkar) [inline modules/mavenDistributionAnalyticsAdapter.js]: also, we should get the response time, since it's available and there's no downside to collecting it

**Comment by partha0solapurkar** [comment]
LGTM based on local testing! I'll leave it up to you whether to get it reviewed for potential JS gotchas from people more familiar with JS 😄

**Review by partha0solapurkar — APPROVED** [review]

**Comment by partha0solapurkar** [comment]
These were the frequencies of the bid status values I saw in half an hour: 
```
Map(
  "nobid" -> 6940189, 
  "error: Bid does not meet price floor" -> 19206, 
  "error: Bid has missing or invalid properties" -> 290, 
  "bid" -> 2436887, 
  "timeout" -> 43071
)
```
Based on this, I don't think it's necessary to condense bid status when recording it in BQ.

**Comment by partha0solapurkar** [comment]
actually, I'll revisit these frequencies later because many users are still running the older (probably cached) code

**Comment by partha0solapurkar** [comment]
By now more than 99% of the users are running the new code. These were the frequencies of the bid status values I saw in an hour:
```
{
  "nobid": 51171942,
  "error: Bid does not meet price floor": 139733,
  "error: Bid has missing or invalid properties": 1188,
  "bid": 23036449,
  "timeout": 490173
}
```
Based on this, I don't think it's necessary to condense bid status when recording it in BQ.

**Comment by martiantim** [comment]
I like these numbers. 0.2% errors with almost all of them being an issue that the vendor probably needs to fix on their end. Good work!

**Pull request merged**